src/api/java/com/minecolonies/api/util/InventoryUtils.java

```diff
@@ -1482,4 +1494,107 @@ public static boolean removeStacksFromProvider(final ICapabilityProvider provide
 
         return true;
     }
+
+    /**
+     * Check if a certain item is in the provider but without the provider being full.
+     * @param provider the provider to check.
+     * @param item the item.
+     * @param itemDamage its damage.
+     * @return the slot or -1.
+     */
+    public static int findFirstSlotInProviderWithNotFull(final ICapabilityProvider provider, final Item item, final int itemDamage)
+    {
+        for (IItemHandler handler : getItemHandlersFromProvider(provider))
+        {
+            int foundSlot = findFirstSlotInItemHandlerWithNotFull(handler, (ItemStack stack) -> compareItems(stack, item, itemDamage));
+            //TODO: When contract is hardened later: Replace this -1 check with a try-catch block.
+            if (foundSlot > -1)
+            {
+                return foundSlot;
+            }
+        }
+
+        return -1;
+    }
+
+    /**
+     * Check if a certain item is in the handler but without the provider being full.
+     * @param handler the handler to check.
+     * @param itemStackSelectionPredicate the selection predicate..
+     * @return the slot or -1.
+     */
+    public static int findFirstSlotInItemHandlerWithNotFull(final IItemHandler handler, @NotNull final Predicate<ItemStack> itemStackSelectionPredicate)
```

Comment:
I think this method Needs to be rewritten.
It should search for a fillable itemstack and return that, but if it cannot find it it should return the first empty stack it finds.
If it can not find the fillable stack and their is no empty Slot then and only then it should return -1.